components/launcher/src/server/mod.rs

```diff
@@ -153,17 +172,136 @@ impl Server {
     }
 
     fn tick(&mut self) -> Result<TickState> {
-        self.reap_zombies();
+        // TODO (CM): Yes, we have `reap_services` as well as
+        // `reap_zombie_orphans`... perhaps they need different
+        // names. However, this is a distinction that might be nice to
+        // collapse in the future.
+        //
+        // `reap_services` is a cross-platform method to reap (and keep
+        // track of) processes that are Habitat
+        // services. `reap_zombie_orphans` is basically a Unix-only
+        // method to take care of any orphan processes that get
+        // re-parented to the Launcher, when it is running as PID 1,
+        // when their parents end before they do.
+        //
+        // There is some natural overlap between the two on Unix
+        // platforms that would be nice to collapse, but it needs to
+        // be done in a way that the basic functionality of process
+        // tracking still works on Windows.
+        self.reap_services();
         match signals::check_for_signal() {
             Some(SignalEvent::Shutdown) => {
                 self.shutdown();
                 return Ok(TickState::Exit(0));
             }
-            Some(SignalEvent::Passthrough(signal)) => self.forward_signal(signal),
+            Some(SignalEvent::WaitForChild) => {
+                // We only return Some if we ended up reaping our
+                // Supervisor; otherwise, we don't need to do anything
+                // special.
+                if let Some(result) = self.reap_zombie_orphans() {
+                    return result;
+                }
+            }
+            Some(SignalEvent::Passthrough(signal)) => {
+                self.forward_signal(signal);
+            }
             None => (),
         }
         self.handle_message()
     }
+
+    /// When the supervisor runs as the init process (e.g. in a
+    /// container), it will become the parent of any processes whose
+    /// parents terminate before they do (as is standard on Linux). We
+    /// need to call `waitpid` on these children to prevent a zombie
+    /// horde from ultimately bringing down the system.
+    ///
+    /// Note that we are not (yet?) doing anything with
+    /// `prctl(PR_SET_CHILD_SUBREAPER, ...)` to make the Launcher a
+    /// subreaper; this behavior currently handles the case when the
+    /// Launcher is running as PID 1.
+    ///
+    /// (See http://man7.org/linux/man-pages/man2/prctl.2.html for
+    /// further information.)
+    #[cfg(unix)]
+    fn reap_zombie_orphans(&mut self) -> Option<Result<TickState>> {
```

Comment:
Yup, this is something I should have caught when I wrote this code. Looks perfect to me!